File: scripts/migrate.mjs
// Chạy: node scripts/migrate.mjs
// Yêu cầu: đã set biến môi trường trong .env.local

import { readFileSync } from "fs";
import { initializeApp, cert } from "firebase-admin/app";
import { getFirestore } from "firebase-admin/firestore";
import { createRequire } from "module";
import { fileURLToPath } from "url";
import { dirname, join } from "path";
import * as dotenv from "dotenv";

const __dirname = dirname(fileURLToPath(import.meta.url));
dotenv.config({ path: join(__dirname, "../.env.local") });

initializeApp({
  credential: cert({
    projectId: process.env.FIREBASE_PROJECT_ID,
    clientEmail: process.env.FIREBASE_CLIENT_EMAIL,
    privateKey: process.env.FIREBASE_PRIVATE_KEY?.replace(/\\n/g, "\n"),
  }),
});

const db = getFirestore();
const data = JSON.parse(
  readFileSync(join(__dirname, "../data/savings.json"), "utf-8")
);

console.log(`Migrating ${data.funds.length} funds...`);

for (const fund of data.funds) {
  const { id, ...rest } = fund;
  await db.collection("funds").doc(id).set({
    ...rest,
    createdAt: Date.now(),
  });
  console.log(`✓ ${fund.title}`);
}

console.log("Done!");
process.exit(0);

File: scripts/migrate.py
import json
import os
from pathlib import Path

# Load .env.local
env_path = Path(__file__).parent.parent / ".env.local"
for line in env_path.read_text(encoding="utf-8").splitlines():
    line = line.strip()
    if line and not line.startswith("#") and "=" in line:
        k, v = line.split("=", 1)
        os.environ.setdefault(k.strip(), v.strip().strip('"'))

import firebase_admin
from firebase_admin import credentials, firestore

cred = credentials.Certificate({
    "type": "service_account",
    "project_id": os.environ["FIREBASE_PROJECT_ID"],
    "client_email": os.environ["FIREBASE_CLIENT_EMAIL"],
    "private_key": os.environ["FIREBASE_PRIVATE_KEY"].replace("\\n", "\n"),
    "token_uri": "https://oauth2.googleapis.com/token",
})

firebase_admin.initialize_app(cred)
db = firestore.client()

data = json.loads(
    (Path(__file__).parent.parent / "data" / "savings.json").read_text(encoding="utf-8")
)

print(f"Migrating {len(data['funds'])} funds...")

for i, fund in enumerate(data["funds"]):
    fund_id = fund["id"]
    payload = {k: v for k, v in fund.items() if k != "id"}
    payload["createdAt"] = i + 1
    db.collection("funds").document(fund_id).set(payload)
    print(f"OK: fund {i+1} uploaded")

print("Done!")
